### PR TITLE
Various edits following a proof read

### PR DESCRIPTION
**Pull request by andrewbonney**
This PR is currently based on top of the client credentials PR so shouldn't be merged until that one is ready.

This is primarily an editorial update after proof reading, but also aims to address a couple of small areas that appeared unclear or unnecessary given how the spec has evolved.

### CODE REVIEW

**Comment by lo-simon** [comment]
I have few points to make.

1. In APIs/AuthorizationAPI.raml
under /{token_endpoint_path}, will it be good to include an example using client_credentials grant?

```
ClientCredentialsGrant:
  grant_type: client_credentials
  client_id: c1068569-80c1-48cf-982e-30c3f930dfe3
  client_assertion_type: urn:ietf:params:oauth:client-assertion-type:jwt-bearer
  client_assertion: eyJhbGciOiJSUzI1NiIsInR5cCI....
```

2. under traits
if we do take on the above example, maybe we should match the client_assertion: example `eyJhbGciOiJFUzI1NiIsImtpZCI6IjE2In0` with the above example `client_assertion: eyJhbGciOiJSUzI1NiIsInR5cCI....` too

3. line 5 of docs/4.4. Behaviour - Access Tokens.md
there is a typo of `Aaccess`

the rest are LGTM

File: docs/4.1. Behaviour - Authorization Servers.md [hunk 1]
```diff
@@ -67,27 +67,26 @@ all OAuth 2.0 registration mechanisms they support.
...
 
 The security considerations outlined in Section 10.1 of [RFC 6749][RFC-6749] SHOULD be abided by with regards to client
 authentication.
 
 ### Client De-Registration
 
-Authorization Servers SHOULD support the de-registration of clients in the case of a compromised client. The means
-by which this client de-registration occurs is beyond the scope of this specification.
+Authorization Servers SHOULD support de-registration of clients in the case of a compromised client. The means by which
+this client de-registration occurs is beyond the scope of this specification.
```
Comment (prince-chrismc) [inline docs/4.1. Behaviour - Authorization Servers.md]: _de-registration_ does not take a hyphen according to google

```suggestion
Authorization Servers SHOULD support unregistering clients in the case of a compromised client. The means by
which the client is unregistered is beyond the scope of this specification.
```

File: docs/4.3. Behaviour - Token Requests.md [hunk 2]
```diff
@@ -18,7 +18,7 @@ The Client Credentials grant MUST be supported by the Authorization Server. This
 confidential clients in machine to machine use cases. Authorization Servers SHOULD provide the means to restrict which
 scopes can be used with this grant type either during dynamic client registration or during token requests.
 
-The Refresh Token Grant MUST be supported in order to ease the refreshing of access tokens and reduce the number user
+The Refresh Token Grant MUST be supported in order to ease the refreshing of Access Tokens and reduce the number user
```
Comment (prince-chrismc) [inline docs/4.3. Behaviour - Token Requests.md]: refresh **must** be supported

File: docs/5.0. Definitions.md [hunk 3]
```diff
@@ -57,10 +57,10 @@ consists of a JSON object in which each key:value pair is referred to as a claim
 A Bearer Token is passed from the Authorization Server to the client after successful authentication of the request.
 Section 5.1. of [RFC 6749][RFC-6749] defines it as consisting of:
 *   an `access_token` - this is the JWT used to access a protected resource.
-*   an `expires_at` value - this coincides with the lifetime of the access token
+*   an `expires_at` value - this coincides with the lifetime of the Access Token
 *   a `token_type` - always "Bearer" within the scope of this document
-*   a `refresh_token` (OPTIONAL) - this is present depending on the grant type. Refresh tokens can be used to retrieve
-further access tokens from the Authorization Server if the shorter-lived access token has expired.
+*   a `refresh_token` (OPTIONAL) - this is present depending on the grant type. Refresh Tokens can be used to retrieve
```
Comment (prince-chrismc) [inline docs/5.0. Definitions.md:62]: here it's marked optional 🤔 the context is unclear in the diff but I think it's mandatory

**Review by prince-chrismc — COMMENTED** [review]
Two tiny comments but it LGTM!

File: docs/4.1. Behaviour - Authorization Servers.md (hunk 1 above)
Comment (garethsb) [reply docs/4.1. Behaviour - Authorization Servers.md]: I agree that deregistration doesn't need a hyphen, but I don't agree that using unregistered is the right fix.
https://english.stackexchange.com/questions/25931/unregister-vs-deregister

File: docs/5.0. Definitions.md (hunk 3 above)
Comment (garethsb) [reply docs/5.0. Definitions.md:62]: If I remember correctly, it's included for the Authorization Code Grant, but not for the Client Credentials Grant?

https://tools.ietf.org/html/rfc6749#section-4.4.3

File: docs/4.3. Behaviour - Token Requests.md (hunk 2 above)
Comment (garethsb) [inline docs/4.3. Behaviour - Token Requests.md]: should that be 'number _of_ user authorization requests'?

File: docs/5.0. Definitions.md (hunk 3 above)
Comment (prince-chrismc) [reply docs/5.0. Definitions.md:62]: Ahh, good point, #77 is not merged yes but yes this should stay

File: docs/4.1. Behaviour - Authorization Servers.md (hunk 1 above)
Comment (prince-chrismc) [reply docs/4.1. Behaviour - Authorization Servers.md]: I read that as well 🤣 I very mildly disagree, perhaps it's cultural but my tendency leans on `un-
https://english.stackexchange.com/a/25933

```suggestion
Authorization Servers SHOULD support deregistering clients in the case of a compromised client. The means by
which the client is deregistered is beyond the scope of this specification.
```

Comment (andrewbonney) [reply docs/4.1. Behaviour - Authorization Servers.md]: I've removed the hyphen for now, but happy to adjust further if desired.

File: docs/5.0. Definitions.md (hunk 3 above)
Comment (andrewbonney) [reply docs/5.0. Definitions.md:62]: I've added some extra language to one of the mentions of access tokens in the other documents as I'd forgotten about this case.

File: docs/4.3. Behaviour - Token Requests.md (hunk 2 above)
Comment (garethsb) [inline docs/4.3. Behaviour - Token Requests.md]: There's a mix through the documentation of whether the word 'grant' is capitalized in phrases like "Refresh Token grant" or "Refresh Token Grant".

File: APIs/AuthorizationAPI.raml [hunk 4]
```diff
@@ -116,13 +110,31 @@ types:
...
+    properties:
+      grant_type:
+        description: The OAuth 2.0 grant type of the request
+        type: string
+        required: true
+      client_id:
+        description: The identifier of the client authenticating
+        type: string
+        required: true
+      client_assertion_type:
```
Comment (dannymeloy) [inline APIs/AuthorizationAPI.raml:123]: Are these the correct keys to be using? In [RFC7521](https://tools.ietf.org/html/rfc7521#section-4.2) and [RFC7523](https://tools.ietf.org/html/rfc7523#section-2.2) it uses these keys for client _authentication_ but lower down we've said that these are the keys required in the body for a token authorization request 

RFC7521 says "Token endpoints can differentiate between assertion-based credentials
   and other client credential types by looking for the presence of the
   "client_assertion" and "client_assertion_type" parameters, which will
   only be present when using assertions for client authentication."

In the examples for Authorization requests in [RFC7521 Section 4](https://tools.ietf.org/html/rfc7521#section-4.1) and [RFC 7523 Section 4](https://tools.ietf.org/html/rfc7523#section-4) it gives `grant_type` and `assertion` as the keys in the body of the request

File: APIs/AuthorizationAPI.raml [hunk 5]
```diff
@@ -190,30 +195,35 @@ types:
...
+        type: TokenAuthorizeRequest | TokenCredentialsRequest | TokenRefreshRequest
         examples:
           AuthCodeGrant:
             grant_type: authorization_code
             code: SplxlOBeZQQYbYS6WxSbIA
             redirect_uri: "http://client.example.com/"
             client_id: ttwCD9hb4OckmcGDnmC3J3Jr
+          ClientCredentialsGrant:
+            grant_type: client_credentials
+            client_id: c1068569-80c1-48cf-982e-30c3f930dfe3
```
Comment (dannymeloy) [inline APIs/AuthorizationAPI.raml:217]: Would be useful to show the client ID and client secret being sent in the Body of the request in the RAML also, as its permitted in RFC6749

File: APIs/AuthorizationAPI.raml (hunk 4 above)
Comment (lo-simon) [reply APIs/AuthorizationAPI.raml:123]: The way how I read the RFCs, there are 2 ways of using the JWT.

1. use for Authorization Grants, doing so the token request will only require the `grant_type` and `assertion`.

2. use for Client Authentication, that require the `client_assertion_type`, `client_assertion` and `client_id`.

For our purpose we are interested using it for case no.2, the Client Authentication.

Another point I would like to make, once client is registered with `"token_endpoint_auth_method": "private_key_jwt"`, client will always require to provide the client_assertion_type, client_assertion (JWT) and client_id while interact with the token endpoint no matter client_credentials or authorization_code grant_type is used. Do you think we need to add that information in our grant examples?

File: APIs/AuthorizationAPI.raml (hunk 5 above)
Comment (lo-simon) [reply APIs/AuthorizationAPI.raml:217]: may be we give different examples, one for using private_key_jwt, and one without.  The one without will require to send client ID and client secret.

File: APIs/AuthorizationAPI.raml (hunk 4 above)
Comment (dannymeloy) [reply APIs/AuthorizationAPI.raml:123]: Thanks for replying to my query Simon. Things were cleared up in [Section 6.2 of RFC7521](https://tools.ietf.org/html/rfc7521#section-6.2) where it explicitly states that the _authentication_ method should be used when the client credentials grant is being used. Perhaps an example of each of the forms that the client credentials grant can take might make things clearer

**Comment by lo-simon** [comment]
some other changes required on those jwks related schemas

```
 APIs/schemas/jwks_response.json            | 45 +--------------------------
 APIs/schemas/jwks_schema.json              | 50 ++++++++++++++++++++++++++++++
 APIs/schemas/register_client_request.json  |  3 +-
 APIs/schemas/register_client_response.json |  3 +-
 4 files changed, 55 insertions(+), 46 deletions(-)

diff --git a/APIs/schemas/jwks_response.json b/APIs/schemas/jwks_response.json
index f34aab8..0959671 100644
--- a/APIs/schemas/jwks_response.json
+++ b/APIs/schemas/jwks_response.json
@@ -3,48 +3,5 @@
   "title": "JWKs Response",
   "description": "JSON Web Key Set to validate Access Token",
   "type": "object",
-  "properties": {
-    "keys": {
-      "description": "The value of the 'keys' parameter is an array of JWK values. By default, the order of the JWK values within the array does not imply an order of preference among them, although applications of JWK Sets can choose to assign a meaning to the order for their purposes, if desired.",
-      "type": "array",
-      "items": {
-        "type": "object",
-        "properties": {
-          "kty": {
-            "type": "string"
-          },
-          "use": {
-            "type": "string"
-          },
-          "key_ops": {
-            "type": "string"
-          },
-          "alg": {
-            "type": "string"
-          },
-          "kid": {
-            "type": "string"
-          },
-          "x5u": {
-            "type": "string",
-            "format": "uri"
-          },
-          "x5c": {
-            "type": "array",
-            "items": {
-              "type": "string"
-            }
-          },
-          "x5t": {
-            "type": "string"
-          },
-          "x5t#S256": {
-            "type": "string"
-          }
-        },
-        "required": ["kty"]
-      }
-    }
-  },
-  "required": ["keys"]
+  "$ref": "jwks_schema.json"
 }
diff --git a/APIs/schemas/jwks_schema.json b/APIs/schemas/jwks_schema.json
new file mode 100644
index 0000000..44804be
--- /dev/null
+++ b/APIs/schemas/jwks_schema.json
@@ -0,0 +1,50 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "title": "JSON Web Key Set",
+  "description": "JSON Web Key Set to validate JSON Web Token",
+  "type": "object",
+  "properties": {
+    "keys": {
+      "description": "The value of the 'keys' parameter is an array of JWK values. By default, the order of the JWK values within the array does not imply an order of preference among them, although applications of JWK Sets can choose to assign a meaning to the order for their purposes, if desired.",
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "kty": {
+            "type": "string"
+          },
+          "use": {
+            "type": "string"
+          },
+          "key_ops": {
+            "type": "string"
+          },
+          "alg": {
+            "type": "string"
+          },
+          "kid": {
+            "type": "string"
+          },
+          "x5u": {
+            "type": "string",
+            "format": "uri"
+          },
+          "x5c": {
+            "type": "array",
+            "items": {
+              "type": "string"
+            }
+          },
+          "x5t": {
+            "type": "string"
+          },
+          "x5t#S256": {
+            "type": "string"
+          }
+        },
+        "required": ["kty"]
+      }
+    }
+  },
+  "required": ["keys"]
+}
diff --git a/APIs/schemas/register_client_request.json b/APIs/schemas/register_client_request.json
index f5a5657..cb7e5df 100644
--- a/APIs/schemas/register_client_request.json
+++ b/APIs/schemas/register_client_request.json
@@ -66,7 +66,8 @@
     },
     "jwks": {
       "description": "Client's JSON Web Key Set document value, which contains the client's public keys",
-      "type": "string"
+      "type": "object",
+      "$ref": "jwks_schema.json"
     },
     "software_id": {
       "description": "A unique identifier string (e.g. a UUID) assigned by the client developer or software publisher",
diff --git a/APIs/schemas/register_client_response.json b/APIs/schemas/register_client_response.json
index 1b3b8ac..c985b6e 100644
--- a/APIs/schemas/register_client_response.json
+++ b/APIs/schemas/register_client_response.json
@@ -85,7 +85,8 @@
     },
     "jwks": {
       "description": "Client's JSON Web Key Set document value, which contains the client's public keys",
-      "type": "string"
+      "type": "object",
+      "$ref": "jwks_schema.json"
     },
     "software_id": {
       "description": "A unique identifier string (e.g. a UUID) assigned by the client developer or software publisher",
```

File: APIs/AuthorizationAPI.raml (hunk 5 above)
Comment (andrewbonney) [reply APIs/AuthorizationAPI.raml:217]: As things stand, the method of sending the client_id and client_secret in the request body isn't explicitly called out by this specification (ie. `client_secret_post`). We could do, but I'm not sure this has an obvious benefit over `client_secret_basic` unless your HTTP stack is very limited, and we're preferring use of `private_key_jwt` anyway.

**Review by garethsb — COMMENTED** [review]
I'm not sure it's good practice to use a `"$ref"` alongside other properties in **Draft 4** since [that part of the spec](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-3) says:
> Any members other than "$ref" in a JSON Reference object SHALL be ignored.

That's still the case in **Draft 7** [here](https://tools.ietf.org/html/draft-handrews-json-schema-01#section-8.3). It becomes valid in **Draft 2019-09** (formerly known as Draft 8) [here](https://tools.ietf.org/html/draft-handrews-json-schema-02#section-8.2.4.1).

So, should we wrap the `$ref` inside an `allOf`, or is it OK as is?

**Review by lo-simon — APPROVED** [review]
LGTM

**Review by garethsb — APPROVED** [review]
LGTM

**Pull request merged**